rfcs/0000-plan-dynanism.md

```diff
@@ -0,0 +1,251 @@
+   (`toDerivedPathsWithHints` in the the nix commands should always be pure functions and not consult the store.)
+
+3. Extend the scheduler and derivation dependencies similarly:
+
+  - Derivations can depended on the outputs of derivations that are themselves derivation outputs.
+    The schedule will substitute derivations to simplify dependencies as computed derivations are built, just like how floating content addressed derivations are realized.
+
+  - Missing derivations get their own full fledged goals so they can be built, not just fetched from substituters.
+
+4. Add a new `outputOf` primop:
+
+   `builtins.outputOf drv outputName` produces a placeholder string with the appropriate string context to access the output of that name produced by that derivation.
+   The placeholder string is quite analogous to that used for floating content addressed derivation outputs.
+   This isn't needed today when we statically know the derivation and thus it's outputs to make attributes for them, but it is needed for built derivations because we don't know either the final derivation or its output names in advanced.
+
+## Deferred import from derivation.
+
+Create a new primop `assumeDerivation` that takes a single expression.
+It's semantics are as follows:
+
+ - If the underlying expression evaluates (shallowly) without needing to build derivations (for `import`, `readFile`, etc.), and is a derivation, simply return that.
+   ```
+   e ⇓ e'
+   e' is derivation
+   ------
+   builtins.assumeDerivation e ⇓ e'
+   ```
+
+- If the underlying expression cannot evaluate (shallowly) without building one or more paths, defer evaluation into a derivation-producing-derivation, and take it's output with `builtins.outputOf`:
+  ```
+  e gets stuck on builds
+  defer = derivation {
+    inputDrvs = builds;
+    buildCommand = "nix-instantiate ${reified e} > $out"
+  }
+  ------
+  builtins.assumeDerivation e ⇓ builtins.outputOf defer "out"
+  ```
+
+This allows downstream non-dynamic derivations to be evaluated without getting stuck on their dynamic upstream ones.
+They just depend on the derivation computed by the deferred eval derivations, and those substitutions are performed by the scheduler.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Here is everything put together with a Haskell and `cabal2nix` example.
+
+Given the following code, and assuming `bar` will depend on `foo`
+```nix
+mkScope (self: {
+  foo = builtins.assumeDerivation (self.callCabal2nix "foo" ./foo);
+  bar = builtins.assumeDerivation (self.callCabal2nix "bar" ./bar);
+})
+```
+We get something like the following derivations with dependencies
+```
+(cabal2nix foo)!out ----> deFoo = deferred eval (fooNix: self.callPackage fooNix)
+(cabal2nix bar)!out ----> deBar = deferred eval (barNix: self.callPackage barNix)
```

Comment:
Where does `deFoo` come from? I'm not quite sure what this means.

Reply:
`deFoo` is the name assigned to the `defferred eval` thing for the next few lines, I'm pretty sure.

Reply:
I think it's only used to disambiguate this text and tell drv apart from deferred drv.

However, I was wondering if this (or something similar) could be made a convention of assumeDerivation to more easily idententify asyncly built derivations.

Reply:
Yeah this example is a mess of unclear notation. (I am disgruntled by this in complete contrast to my standing by the sequent notation I like a lot.)